File: src/hw2/quicksort/Quicksort.scala
object Quicksort {
  def swap(a: Array[Int], pos1: Int, pos2: Int): Unit = {
    val stash = a(pos1)
    a(pos1) = a(pos2)
    a(pos2) = stash
  }

  def quicksort(a: Array[Int], low: Int, hi: Int): Unit = {
    if (low < hi) {
      val p = partition(a, low, hi)
      quicksort(a, low, p-1)
      quicksort(a, p+1, hi)
    }
  }

  def partition(subArray: Array[Int], low: Int, hi: Int): Int = {
    val pivot = hi;
    var i = low;
    for (
      j <- low to hi
      if subArray(j) < subArray(pivot)
    ) {swap(subArray, i, j); i+=1}

    swap(subArray, i, pivot);
    return i
  }
}

File: src/hw2/quicksort/quicksort.c
#include <stdio.h>

int (int[] array, int left, int right) {
    int wall = left;
    int pivot = right;

    for (int i = left; i < right; i++) {
        if (array[i] <= array[pivot]) {
            int temp = array[wall];
            array[wall] = array[i];
            array[i] = temp;
            wall++;
        }
    }
    int temp = array[wall];
    array[wall] = array[pivot];
    array[pivot] = temp;

    return wall;
}

void quicksort(int[] array, left, right) {
    if (left >= right) {
        return;
    }

    int pivot = partition(array, left, right);
    quicksort(array, left, pivot - 1);
    quicksort(array, pivot + 1, right);
}


int main(void) {
    printf("hello world\n");
    return 0;
}

